capstone/src/test/java/com/google/sps/AssociationDatastoreTest.java

```diff
@@ -0,0 +1,153 @@
+    addComment(datastore, "test", "94124", false);
+    addComment(datastore, "testing", "94122", true);
+
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    ArrayList<AssociationInput> input = association.getComments();
+    assertEquals(
+        input,
+        new ArrayList<AssociationInput>(
+            Arrays.asList(
+                new AssociationInput(
+                    "test", new ArrayList<String>(Arrays.asList("Bayview", "SF"))))));
+  }
+
+  public boolean isValidAssociationDatastore(ArrayList<AssociationResult> results) {
+    HashSet<String> seen = new HashSet<String>();
+    for (AssociationResult result : results) {
+      if (seen.contains(result.getContent())) {
+        return false;
+      }
+      seen.add(result.getContent());
+    }
+    return true;
+  }
+
+  @Test
+  public void addThenRetrieveTest() {
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "SF");
+    ArrayList<AssociationResult> retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
+    assertEquals(data, retrieved);
+  }
+
+  @Test
+  public void addThenRetrieveDifferentScopeTest() {
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "Bayview");
+    ArrayList<AssociationResult> retrieved = association.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
+    assertEquals(new ArrayList<AssociationResult>(), retrieved);
+  }
+
+  @Test
+  public void dataPersistenceTest() {
+    ArrayList<AssociationResult> data =
+        new ArrayList<AssociationResult>(
+            Arrays.asList(
+                new AssociationResult("hi", 12.4f, 118.0f, false),
+                new AssociationResult("test1", 1.3f, 123.4f, true)));
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    AssociationDatastore association = new AssociationDatastore(datastore);
+    association.storeResults(data, "SF");
+    AssociationDatastore newAssociation = new AssociationDatastore(datastore);
+    ArrayList<AssociationResult> retrieved = newAssociation.loadPreviousResults("SF");
+    assert (isValidAssociationDatastore(retrieved));
```

Comment:
Do we need these extra parens? For example, can we do `assert isValidAssociationDatastore(...)`. Additionally, should we add error strings to the `assert`s via `assert False : "Some error message"`.

Reply:
Moved it to assertTrue with error messages!